airbyte-integrations/connectors/destination-bigquery/src/main/kotlin/io/airbyte/integrations/destination/bigquery/BigQueryDestination.kt

```diff
@@ -0,0 +1,394 @@
+            val bigquery = getBigQuery(config)
+            val uploadingMethod = getLoadingMethod(config)
+
+            checkHasCreateAndDeleteDatasetRole(bigquery, datasetId, datasetLocation)
+
+            val dataset = getOrCreateDataset(bigquery, datasetId, datasetLocation)
+            if (dataset.location != datasetLocation) {
+                throw ConfigErrorException(
+                    "Actual dataset location doesn't match to location from config",
+                )
+            }
+            val queryConfig =
+                QueryJobConfiguration.newBuilder(
+                        String.format(
+                            "SELECT * FROM `%s.INFORMATION_SCHEMA.TABLES` LIMIT 1;",
+                            datasetId,
+                        ),
+                    )
+                    .setUseLegacySql(false)
+                    .build()
+
+            if (UploadingMethod.GCS == uploadingMethod) {
+                val status = checkGcsPermission(config)
+                if (status!!.status != AirbyteConnectionStatus.Status.SUCCEEDED) {
+                    return status
+                }
+            }
+
+            val result = executeQuery(bigquery, queryConfig)
+            if (result.getLeft() != null) {
+                return AirbyteConnectionStatus()
+                    .withStatus(AirbyteConnectionStatus.Status.SUCCEEDED)
+            } else {
+                throw ConfigErrorException(result.getRight())
+            }
+        } catch (e: Exception) {
+            log.error(e) { "Check failed." }
+            throw ConfigErrorException((if (e.message != null) e.message else e.toString())!!)
+        }
+    }
+
+    /**
+     * This method does two checks: 1) permissions related to the bucket, and 2) the ability to
+     * create and delete an actual file. The latter is important because even if the service account
+     * may have the proper permissions, the HMAC keys can only be verified by running the actual GCS
+     * check.
+     */
+    private fun checkGcsPermission(config: JsonNode): AirbyteConnectionStatus? {
+        val loadingMethod = config[bqConstants.LOADING_METHOD]
+        val bucketName = loadingMethod[bqConstants.GCS_BUCKET_NAME].asText()
+        val missingPermissions: MutableList<String> = ArrayList()
+
+        try {
+            val credentials = getServiceAccountCredentials(config)
+            val storage =
+                StorageOptions.newBuilder()
+                    .setProjectId(config[bqConstants.CONFIG_PROJECT_ID].asText())
+                    .setCredentials(credentials)
+                    .setHeaderProvider(getHeaderProvider())
+                    .build()
+                    .service
+            val permissionsCheckStatusList =
+                storage.testIamPermissions(bucketName, REQUIRED_PERMISSIONS)
+
+            missingPermissions.addAll(
+                StreamUtils.zipWithIndex(permissionsCheckStatusList.stream())
```

Comment:
does `permissionsCheckStatuslist.withIndex` work here? This function is an example of a place where types are far from obvious and should be set explicitely.
I'd rather convert the returned list into a `Map<Permission, IsEnabled>` than going index-based: `REQUIRED_PERMISSIONS.zip(permissionsCheckStatusList).toMap()`. Then, missingPermissions becomes `map.filterValues`

Reply:
Yup, see above comment about `StreamUtils`